docs/content/v0.20/architecture.md

```diff
@@ -1,36 +1,479 @@
+to understand).
+
+<figure>
+  <img src="assets/placement-controller.png"  alt="Placement Controller">
+  <figcaption align="center">Figure 3 - Placement Controller</figcaption>
+</figure>
+
+At startup, the controller code sets up the dynamic informers, the event
+handler and the work queue as follows:
+
+- lists all API resources
+- Filters out some resources
+- For each resource:
+  - Creates GVK key
+  - Registers Event Handler
+  - Starts Informer
+  - Indexes informer and lister in a map by GVK key
+- Waits for all caches to sync
+- Starts N workers to process work queue
+
+The reflector is started as part of the informer and watches specific
+resources on the WDS API Server; on create/update/delete object events it
+puts a copy of the object into the local cache. The informer invokes the
+event handler. The handler implements the event handling functions
+(AddFunc, UpdateFunc, DeleteFunc)
+
+A typical event flow for a create/update object event will run as
+follows:
+
+1.  Informer invokes the event handler AddFunc or UpdateFunc
+
+2.  The event handler does some filtering (for example, to ignore update
+    events where the object content is not modified) and then creates a
+    key to store a reference to the object in the work queue. The key
+    contains the GVK key used to retrieve a reference to the informer
+    and lister for that kind of object, and a namespace/name key to
+    retrieve the actual object. Storing the key in the work queue is a
+    common pattern in client-go as the object may have changed in the
+    cache (which is kept updated by the reflector) by the time a worker
+    gets a copy from the work queue. Workers should always receive the
+    key and use it to retrieve the object from the cache.
+
+3.  A worker pulls a key from the work queue, and then does the
+    following processing:
+
+    -  Uses the GVK key to get the reference to the lister for the
+        object
+    -  Gets the object from the lister cache using the key that was
+        generated with the MetaNamespaceKeyFunc.
+    -  Gets the lister for Placement objects and list all placements
+    -  Iterates on all placements, and for each placement:
+        - evaluates the placement label selector vs. object labels to
+          check if there is a match.
+        - If there is a match, uses the lister for ManagedCluster and
+          find the matching clusters using the label selector expression
+          for clusters.
+        - If there are matching clusters, add the names of the cluster
+          to a hashmap setting the name of the cluster as a key. Cluster
+          groups from different placements are merged together.
+    -  If there are no matching clusters, the worker returns without
+        actions and is ready to process other events from the queue.
+    -  If there are matching clusters:
+        - Wraps the object into a ManifestWork
+        - Adds a label to the ManifestWork that is used to track the
```

Comment:
I find it confusing to start by talking about _the_ label and later recognize that there is a label per (Placement, cluster). I think it would be better to start off by explaining that there is a label per (Placement, cluster).

Reply:
fixed